Query Filters

amercader opened the pull request.

amercader committed:
First draft and tests of filters validator

amercader committed:
Implement more top level checks

amercader committed:
First draft of the filters spec

amercader committed:
Merge branch 'main' into filters

amercader committed:
Merge branch 'main' into filters

amercader commented on doc/filters.md:
```diff
@@ -0,0 +1,188 @@
+
+TODO: Additional operators can be added by plugins
+
+## Top level operators
```
@wardi consider these filters:

```
"filters": {
    "field1": "value1",
    "$or": [
        {"field2": ["value21", "value22"]},
        {"field3": {"gte": 9000, "lt": 1000}}
    ]
}
```

Would this be how you'd expect them to be parsed?

```
filters = FilterOp(op='$and', field=None, value=[
    FilterOp(op='eq', field='field1', value='value1'),
    FilterOp(op='$or', field='None', value=[
        FilterOp(op='in', field='field2', value=['value21', 'value22']),
        FilterOp(op='$and', field=None, value=[
            FilterOp(op='gte', field='field3', value=9000),
            FilterOp(op='lt', field='field3', value=10000),            
        ])
    ])
])
```

If so perhaps we need to perhaps rename "Top level operators" to "Logial operators" as they can be nested and not just at the top level?

wardi replied:
That parsing looks correct.

Naming things is so hard. I suggested "top level" because operators are normally after fields, "logical operators" is good but I can see sites wanting to extend them, so really we need a name for:

- things that can appear as a filter, e.g.
  - `"fieldname": ...`
  - `"$op": ...`
- things that can appear after a field name is given, e.g.
  - `"string"`
  - `42`
  - `true`
  - `["some, "values"]`
  - `{"op": ..., ... }`
  - `[{"op": ..., ... }, "or", "values", ...]`
 
how about **"filter operations"** and **"field operations"**?

- "$op" is a **filter operator** and "op" is a **field operator**
- "filters" values are either a filter operations dict or a list of filter operations dicts

amercader replied:
I'm not in love with it, but I don't have any better alternative yet :) 
I'll have another thought but we can go with this for now.

wardi commented on doc/filters.md:
````diff
@@ -0,0 +1,188 @@
+For example:
+
+```
+"filters": {
+    "field1": "value",
+    "$or": [
+        {"field2": ["value1", "value2"]},
+        {"field3": {"gte": 9000, "lt": 1000}}
+    ]
+}
````
If we use real looking examples it could make the docs more engaging:

```suggestion
"filters": {
    "incident": "noise complaint",
    "$or": {
        "resolution": ["unresolved", "in progress"],
        "year": {"gt": 2024}
    },
    "sector": [2, 8, 10, {"gte":15, "lte": 30}, 35],
}
```

I like this as a first example because it covers most of the functionality we'll describe: a short-form "eq" filter operation, a logical filter operation, a short-form "in" field operation, a simple "gt" field operation, and a more complex but understandable combined logical "or" and "in" field operation

wardi commented on doc/filters.md:
````diff
@@ -0,0 +1,188 @@
+Top level operators must always contain lists of at least two filters, i.e. all of the 
+following are incorrect:
+
+```
+{
+  "$or": {"field1": "value1"},
+}
+
+{
+  "$and": [
+    {"field1": "value1"}
+  ]
+}
````
We can allow 0- and 1-element `"$or"` and `"$and"` values if we treat them like the python `any()` and `all()` functions.

```python
>>> any([])  # same as "$or": []
False
>>> all([])  # same as "$and": []
True
```
The benefit is when search filters for API calls are generated by code that code doesn't need to special-case the 0 and 1-element cases.

amercader replied:
I can see how ignoring empty lists is useful, but I don't understand how considering them False or True works with the actual parsing, ?

For example, what does this parse to?

```
{
   "field1": "value1",
   "$or": []
}
```
Is it something like:
```
field1 = value AND False
```

wardi replied:
Yes `false` works. We'd still generate all the FilterOps and the back end would generate something like `where "field1" = 'value1' and false` then return no records for us.

amercader replied:
If the goal is make it easier for code generated filters by not chocking on empty lists so users don't have to worry about them, getting no results because you inadvertently sent  an empty list might be more confusing. You effectively need to check that you are not sending empty lists to get some results back, which defeats the purpose of not special casing empty lists right?

wardi replied:
Getting nothing back from an empty `$or` list is consistent and shouldn't cause an error.
 - `"color": ["red", "blue"]` might give some results
 - `"color": []` we would expect to give 0 results

Which is the same as:
- `"$or": [{"color": "red"}, {"color": "blue"}]` might give some results
- `"$or": []` we would expect to give 0 results

Getting 0 results is also fine because this might be a smaller part of a larger `$or` condition that we don't want to error out just because one part gives no results.

wardi replied:
I like the idea of treating `$and` and `$or` like `all()` and `any()` because I think it's a better interface, but if you think it causes confusion then we could raise an error in the 1-element and 0-element cases instead.

wardi commented on doc/filters.md:
````diff
@@ -0,0 +1,188 @@
+
+```
+  {
+    "<field_name>": {"<operator1>" : <value1>}, {"<operator2>" : <value2>}
````
```suggestion
    "last_name": {"gte" : "Jeong", "lte": "Romano"}
```

wardi commented on doc/filters.md:
````diff
@@ -0,0 +1,188 @@
+is equivalent to:
+
+```
+    {
+      "$and": [
+        "<field_name>": {"<operator1>" : <value1>},
+        "<field_name>": {"<operator2>" : <value2>}
+      ]
+    }
````
```suggestion
    {
      "$and": [
        {"last_name": {"gte" : "Jeong"}},
        {"last_name": {"lte" : "Romano"}}
      ]
    }
```

wardi commented on doc/filters.md:
````diff
@@ -0,0 +1,188 @@
+
+```
+
+If filters contain lists of operator/value dictionaries, these are combined with the OR operator, e.g.:
````
```suggestion
If a field operation is a lists of values or other field operations, the values are grouped into an "in" field operation and the field operations are ORed together, e.g.:
```

wardi commented on doc/filters.md:
````diff
@@ -0,0 +1,188 @@
+If filters contain lists of operator/value dictionaries, these are combined with the OR operator, e.g.:
+
+```
+  {
+    "<field_name>": [
+      {"<operator1>" : <value1>}, 
+      {"<operator2>" : <value2>}
+    ]
+  }
````
```suggestion
  {
    "year": [1990, 2010, {"gt", 2023}]
  }
```

wardi commented on doc/filters.md:
````diff
@@ -0,0 +1,188 @@
+is equivalent to:
+
+```
+  {
+    "$or": [
+      {"field_name": {"<operator1>" : <value1>}}, 
+      {"field_name": {"<operator2>" : <value2>}}
+    ]
+  }
````
```suggestion
  {
    "$or": [
      {"year": {"in": [1990, 2010]}}, 
      {"year": {"gt": 2023}}
    ]
  }
```

wardi commented on doc/filters.md:
```diff
@@ -0,0 +1,188 @@
+{
+  "$and": ["value1", "value2", "value3"]
+}
+
```
````suggestion

When a dict is given as the value of an `$or` filter operation each key is treated as a separate part of the OR operation:

```
"filters": {
  "$or": {
    "color": "red",
    "size": 10
  }
}
```

is equivalent to:
```
"filters": {
  "$or": [
    {"color": "red"},
    {"size": 10}
  ]
}
```

This rule does **not** extend to values within a list, e.g.
```
"filters": {
  "$or": [
    {"color": "yellow", "size": 9},
  ]
}
```

is equivalent to:
```
"filters": {
  "color": "yellow", "size": 9
}
```

````

wardi replied:
@amercader what do you think about this one? On the one hand it allows users to skip a bunch of extra dicts in their `$or` operations, but on the other hand someone might eventually try to `"$or": {"score": {"gt": 90}, "score": {"lt": 20}}` and our code will **silently ignore** one of the conditions.

amercader replied:
I think we need to draw a line somewhere and avoid complicating the spec and the implementations, specially if it can lead to edge cases like the one you mention. I quite like enforcing lists of dicts for the filter operators. Let's test ride this and we can always extend it later.

amercader committed:
Improve spec examples

Following @wardi's suggestions

Co-authored-by: Ian Ward <ian@excess.org>

wardi commented on doc/filters.md:
````diff
@@ -0,0 +1,186 @@
+```
+"filters": {
+    "incident": "noise complaint",
+    "$or": {
+        "resolution": ["unresolved", "in progress"],
+        "year": {"gt": 2024}
+    },
````
```suggestion
    "$or": [
        {"resolution": ["unresolved", "in progress"]},
        {"year": {"gt": 2024}}
    ],
```
as per discussion below

wardi commented on doc/filters.md:
````diff
@@ -0,0 +1,186 @@
+
+```
+
+Note that fields starting with the character `$` must be escaped with another `$` character:
+
+```
+filters: {
+    "$$some_field": "some_value"
+}
+```
````
````suggestion
Note that fields starting with the character `$` must be escaped with another `$` character, for example a `$AU` field would be given as:

```
filters: {
    "$$AU": {"lt": 150}
}
```
````

amercader committed:
Update examples and language in spec

Co-authored-by: Ian Ward <ian@excess.org>

amercader committed:
First pass at the validator based on updated spec

Still just returning dicts, not FilterOp objects

amercader committed:
Add repr method to FilterOp to make easier to debug

amercader committed:
Translate normalized filters dict to FilterOp objects

amercader committed:
Adapt tests

amercader commented:
@wardi See the latest version for a first pass at an implementation for the validator/normalizer. It still needs a few more checks and once I add all tests to cover the different combinations it will be easier to follow, but some quick tests I was doing look promising:

https://gist.github.com/amercader/a49ef44e2f5758781a7d6d8d44af2f8a

This is just to see how you feel about the overall approach or you think we should go another route

wardi commented on ckanext/search/filters.py:
```diff
@@ -4,51 +4,189 @@
+
+        return (
+            f"FilterOp(op={repr(self.op)}, field={repr(self.field)}, value={value_str})"
+        )
```
Nice

wardi commented on ckanext/search/filters.py:
```diff
@@ -4,51 +4,189 @@
+
+    out = []
+    for k, v in filters_dict.items():
+        if k in FILTER_OPERATORS:
```
for consistency with above (and to allow custom filter operators in the future)
```suggestion
        if k.startswith("$") and not k.startswith("$$"):
```

wardi commented on ckanext/search/filters.py:
```diff
@@ -4,51 +4,189 @@
+            filters[op] = _process_filter_operator_members(filters[op])
 
+    # Translate output dict to immutable FilterOp
+    out = _to_filter_op(filters)
```
Is there a reason to do this as a separate recursive step instead of as part of the processing/validation?

wardi commented:
It's nicer if the validator can return multiple errors for the field at once. Maybe return `(filters, errors list)` from the recursive functions and when there are errors a placeholder could be used as the filters since the value won't be used.

amercader committed:
Tests for top level operators

amercader committed:
Refactor to return FilterOp during validation/processing

amercader committed:
Combine nested filter operators of the same type

For logical ANDs and ORs, merge all nested filters of the same type to
simplify the resulting FilterOp objects, eg given:

    {
        "field1": "value1",
        "$and": [
            {"field2": {"gte": 5, "lte": 7}},
            {
                "$and": [
                    {"field3": "value3"},
                    {"field4": "value4"},
                ]
            }
        ]
    }

Return this:

    FilterOp(
        field=None,
        op="$and",
        value=[
            FilterOp(field="field1", op="eq", value="value1"),
            FilterOp(field="field2", op="gte", value=5),
            FilterOp(field="field2", op="lte", value=7),
            FilterOp(field="field3", op="eq", value="value3"),
            FilterOp(field="field4", op="eq", value="value4"),
        ]
    )

Instead of this:

    FilterOp(
        field=None,
        op="$and",
        value=[
            FilterOp(field="field1", op="eq", value="value1"),
            FilterOp(
                field=None,
                op="$and",
                value=[
                    FilterOp(
                        field=None,
                        op="$and",
                        value=[
                            FilterOp(field="field2", op="gte", value=5),
                            FilterOp(field="field2", op="lte", value=7),
                        ]
                    ),
                    FilterOp(
                        field=None,
                        op="$and",
                        value=[
                            FilterOp(field="field3", op="eq", value="value3"),
                            FilterOp(field="field4", op="eq", value="value4"),
                        ]
                    )
                ]
            )
        ]
    )

amercader commented:
@wardi some questions:

* I haven't found an elegant solution for the multiple errors, will keep thinking about it
* What do you think of https://github.com/amercader/ckanext-search/pull/1/commits/457722e2f15955cd84f3ae94c33f93dc1f2db5e9? It complicates a bit the code but it simplifies a lot the returned FilterOps once you start nesting operators, and at least for ANDs and ORs the laws of logic tell us that it should be fine (`A ∧ (B ∧ C) ≡ A ∧ B ∧ C`)
* Related to our discussion on how to treat empty $or / $and, what do you think should be returned here?
```
     filters = {
         "$or": [
             {"field1": "value1"},
         ]
     }
```
* The only main thing missing is some sort of validation on the field operators (e.g. a list of built-in operators that can be extended by pluginsparameters depending on the operator present. I'm thinking a mini-schema of operators/keys: list of validators for the params so we can call navl.validate on them, something like:

```
{
    "gt": [string_or_number]
    "in": [list_with_strings_or_numbers]
}
```
* Also thinkig about built-in protections, maybe limiting the recursion limit and/or enforcing a hard limit on number of filters?

wardi commented on ckanext/search/filters.py:
```diff
@@ -55,7 +62,56 @@ def _check_filter_operator(key: str, value: Any) -> None:
+
+    combined_filters = []
+    for filter_ in child_filters:
+        if filter_.op == parent_operator and parent_operator in COMBINE_OPERATORS:
```
Is "commutative" the right term for this kind of operation? When plugins register operators we can let them declare if these new operators can be combined/flattened in the same way too.

amercader replied:
If I remember correctly from my school days commutative relates to the order of operands (e.g. A + B = B + A)

wardi replied:
what we're expressing is "(A + B) + C = A + (B + C)", is there a different word for that?

amercader replied:
[*Well, acchckhually...*](https://c.tenor.com/nnUyxuqaeNcAAAAd/tenor.gif) what we are describing is (A+B) + C = A + B + C (or rather (A ∧ B) ∧ C = A ∧ B ∧ C  for AND operations). Is it something like [associatiative](https://en.wikipedia.org/wiki/Associative_property) what we mean?

wardi commented on ckanext/search/tests/test_filters.py:
```diff
@@ -125,6 +234,77 @@ def test_filters_list_of_filters_combined_as_or():
+
 # TODO: what do we expect here?
 # def test_filters_single_or():
 #     filters = {
```
Single-value or/and are either the same as unwrapping the inner operator (behaving like `all()`/`any()`) or an error if we want to be strict with this case too.

wardi commented:
> @wardi some questions:
> 
>     * I haven't found an elegant solution for the multiple errors, will keep thinking about it

Each recursive call can return a `(filter_op, error_list)` tuple. For `$and` and `$or` the children's `error_list`s are combined and returned as its `error_list` At the top level when the `error_list` isn't empty we return that as the `filters` validation errors.

This wouldn't have path/nesting information to each error but if the error text includes part of the values that caused the error then users should be able to figure out where the problem lies.

>     * The only main thing missing is some sort of validation on the field operators (e.g. a list of built-in operators that can be extended by pluginsparameters depending on the operator present. I'm thinking a mini-schema of operators/keys: list of validators for the params so we can call navl.validate on them, something like:
> 
> 
> ```
> {
>     "gt": [string_or_number]
>     "in": [list_with_strings_or_numbers]
> }
> ```

The types allowed will depend on the types of the underlying fields (e.g. date fields can also be ordered, "in" could work on almost anything) so the operators will need a `validate_parameter(filter_op, search_schema)` method that can do parameter validation based on the search schema, field name and values passed. navl doesn't fit nicely here.

>     * Also thinkig about built-in protections, maybe limiting the recursion limit and/or enforcing a hard limit on number of filters?

A hard limit on the number of filter ops sounds good. That would solve the recursion issue as long as we pass in the remaining limit to the recursive function.

amercader committed:
Implement multiple error reporting

amercader committed:
Small refactor

amercader committed:
Field operator validation, return multiple errors

The validator now needs a search schema to be able to perform field
based validation. For instance in metadata search that will be the new
search combined search schema (ckanext.search.schema) and for the
datastore filters the schema coming from datastore_info

wardi commented on ckanext/search/filters.py:
```diff
@@ -45,6 +45,17 @@ def __repr__(self) -> str:
         )
 
 
+def _validate_field_operator(field_name, operator, value, search_schema):
```
if our search_schema was a dict with field names as keys we wouldn't need this function

amercader replied:
I used the form `{"fields": [{"name": "field1"}]` because that's the schema format I've been using so far in the wider search implementation (i.e. [this one](https://github.com/amercader/ckanext-search/blob/70c6574c8c9a0394a796ec186fa7b270be9a42e6/ckanext/search/schema.py#L46)). This in turn was inspired by the scheming schemas and I think it's more flexible if we want to add new keys in the future like "default_sorting" or whatever.
Regardless of the format of the schema, the validation needs to be performed anyway for each provided filter so it makes sense to have it in a separate function? We might have to add plugin hooks in there to add new operators as well.
Unless I misunderstood!

wardi replied:
Right, having other keys makes sense so I should have said the "fields" value could be a dict with field names as keys.

scheming schemas use a list of fields because the order of fields matters for displaying in the form and for displaying extra metadata. Many places in scheming I wish the fields were in a dict to simplify the code (and now python dicts are ordered so the list isn't required)

Here we don't really care about the order of fields so we could do something different than scheming to simplify working with search schemas. Even when a search schema is generated from a scheming schema the field names can be different e.g. because they need to be flattened, so it's not like we can reuse the schema directly.

This is just an idea, and there might be some other problem I'm not thinking of.

amercader replied:
I implemented names-as-keys schemas in [59607be](https://github.com/amercader/ckanext-search/pull/1/commits/59607be344995c835479108397fa77d21087aa7d).

amercader committed:
Some more tests

wardi commented on ckanext/search/filters.py:
```diff
@@ -115,66 +128,80 @@ def _combine_filter_operator_members(
-            return None, errors
-
-        key = list(item.keys())[0]
+            item_errors.append(f"Filter operations can only have one key: {item}")
```
do we have this restriction? Aren't all operations "AND"ed together?

amercader replied:
I removed this restriction in [b692f95](https://github.com/amercader/ckanext-search/pull/1/commits/b692f951b7995b8e7c733a18e48e8ffa10c3b5fc) (at the expense of some more complexity :yin_yang: )

wardi commented on ckanext/search/filters.py:
```diff
@@ -45,6 +45,17 @@ def __repr__(self) -> str:
 
+def _validate_field_operator(field_name, operator, value, search_schema):
+
+    # Check field exists
+    field = [f for f in search_schema.get("fields", {}) if f["name"] == field_name]
+    if not field:
+        return [f"Unknown field: {field_name}"]
+
+    field = field[0]
+    pass
```
```suggestion
    if all(f['name'] != field_name for f in search_schema['fields']):
        return [f"Unknown field: {field_name}"]
```

wardi commented on ckanext/search/filters.py:
```diff
@@ -115,66 +128,80 @@ def _combine_filter_operator_members(
-                continue
+        if item_errors:
+            errors.extend(item_errors)
+        else:
```
continue instead of else?

wardi commented on ckanext/search/tests/test_filters.py:
```diff
@@ -39,65 +54,66 @@ def test_filters_unknown_top_operators():
 
-    result = query_filters_validator(filters)
+    result = query_filters_validator(filters, search_schema)
     assert result and result.field == "$some_field" and result.op == "eq"
```
If you prefer this style should also work and might give a clearer traceback on failure
```suggestion
    assert result == FilterOp("$some_field", "eq", "some_value")
```

wardi replied:
oh, I see you're already doing this below, my bad

wardi commented on ckanext/search/filters.py:
```diff
@@ -115,66 +128,80 @@ def _combine_filter_operator_members(
+        item_errors = []
         if not isinstance(item, dict):
-            errors.append(f"Filter operation members must be dictionaries: {item}")
+            item_errors.append(f"Filter operation members must be dictionaries: {item}")
```
Since the error is about the type maybe the repr is better?
```suggestion
            item_errors.append(f"Filter operation members must be dictionaries: {item!r}")
```
or even `{json.dumps(item)}`  so it looks right for API users?

wardi commented on ckanext/search/tests/test_filters.py:
```diff
@@ -317,12 +333,69 @@ def test_filters_list_of_filters_with_or_operator():
+    with pytest.raises(ValidationError) as e:
+        query_filters_validator(filters, default_search_schema)
+
+    assert e.value.error_dict["filters"][0] == "Unknown field: random_field"
```
I tend to write these like this so failures don't turn into errors
```suggestion
    assert e.value.error_dict == {'filters':["Unknown field: random_field"]}
```

amercader committed:
Improve test based on review

amercader committed:
Remove restriction on number of filter operation keys

On filter operation members, see test included in the commit

amercader committed:
Migrate fields in search schema to names as keys

i.e

```
{
    "fields" : {
        "field1": {"type": "text"},
        "field2": {"type": "text"}
    }
}
```
rather than

```
{
    "fields" : [
        {"field1": {"type": "text"}},
        {"field2": {"type": "text"}}
    ]
}

See discussion here:
https://github.com/amercader/ckanext-search/pull/1#discussion_r2109125631

```

amercader committed:
Hard limit on nested operators

amercader committed:
Hard limit on total number of operations

Rather than pass a current number of operations around which was hard to
track, or check the total filter ops at the end, which might mean wasted
processing, we check the number of child ops at different points of processing.

Added a new op_count() method to FilterOp to aid with that.

amercader committed:
Rename filter parsing function

As it is not really a CKAN validator

amercader committed:
Hook filter parser to search schema

Create an actual CKAN validator that calls the parsing function and
hooks with the navl validation. Update logic and tests regarding filters
that used plain dictionaries to use FilterOp objects

TODO: tests for perm labels handling, Solr perm labels filter

amercader committed:
Use closure to check max filterop number

amercader commented:
> another option is defining the recursive functions in the same function as the new_filter_op closure

I'm not sure I like this. At this point perhaps it's easier to implement it as a `FilterParser` class and just use internal properties to keep errors/op counts?. We can still have a single `parse_query_filters()` method as the external API:

```python
  def parse_query_filters(
      input_value: Optional[Union[Dict[str, Any], List[Dict[str, Any]]]],
      search_schema: dict,
  ) -> Optional[FilterOp]:

    parser = FilterParser(search_schema)
   
    filters, errors = parser.parse(input_value)

      if errors:
          raise ValidationError({"filters": errors})

      return filters

```

wardi commented:
if the counts and errors are tracked on the class you might as well do the parsing in `__init__`

amercader committed:
Implement parsing as a FiltersParser class

This simplifies the functions signatures as we avoid passing filterop
factories around.
Parsing is done in the `__init__` method to avoid thread safety issues.
The module API remains the same, the `parse_query_filters()` function.

amercader commented:
thanks for the feedback @wardi, I think the code looks much nicer now. It's not trivial to understand but not too difficult considering all that we are doing. I'll merge this and work on integrating it in the Solr and ES providers, which might surface issues with the spec or parser but that can be dealt with in another PR

wardi commented on ckanext/search/filters.py:
```diff
@@ -61,365 +61,400 @@ def __repr__(self) -> str:
 
-def _is_dict_or_list_of_dicts(value: Any) -> bool:
-    return isinstance(value, dict) or _is_list_of_dicts(value)
+        self.filters, self.errors = self._parse_query_filters(
```
You have a class now so your `_parse_query_filters` and other methods could return only filters and when errors are encountered they would be added directly to `self.errors` instead.

amercader committed:
Simplify code handling errors and search schema

Use internal class references instead of passing errors and
search_schema around
